File: yandex_back_prep_python/leetcode/443. String Compression.py
class Solution:
    def compress(self, chars: List[str]) -> int:
        l, r, r_pos = 0, 0, 0

        while r < len(chars):
            while r < len(chars) - 1 and chars[r] == chars[r + 1]:
                r += 1

            str_len = r - l + 1

            chars[r_pos] = chars[l]
            r_pos += 1

            if str_len > 1:
                for c in str(str_len):
                    chars[r_pos] = c
                    r_pos += 1

            r += 1
            l = r

        return r_pos

File: yandex_back_prep_python/leetcode_cpp/443._String_Compression.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int compress(vector<char>& chars) {
        
        int l = 0;
        int r = 0;
        while (r < chars.size()) {
            int _l = r;
            while (r < chars.size()-1 && chars[r] == chars[r+1]) {
                r +=1;
            }

            int ln = r-_l+1;
            chars[l] = chars[r];
            l +=1;

            if (ln > 1) {
                for (const char& c : to_string(ln)) {
                    chars[l] = static_cast<int>(c);
                    l +=1;
                }
            } 

            r+=1;

        }

        return l;
    }


};
